src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleBinaryOperator.cs

```diff
@@ -0,0 +1,355 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed partial class LocalRewriter
+    {
+        /// <summary>
+        /// Rewrite `GetTuple() == (1, 2)` to `tuple.Item1 == 1 &amp;&amp; tuple.Item2 == 2`.
+        /// Also supports the != operator, nullable and nested tuples.
+        ///
+        /// Note that all the side-effects for visible expressions are evaluated first and from left to right. The initialization phase
+        /// contains side-effects for:
+        /// - single elements in tuple literals, like `a` in `(a, ...) == (...)` for example
+        /// - nested expressions that aren't tuple literals, like `GetTuple()` in `(..., GetTuple()) == (..., (..., ...))`
+        /// On the other hand, `Item1` and `Item2` of `GetTuple()` are not saved as part of the initialization phase of `GetTuple() == (..., ...)`
+        ///
+        /// Element-wise conversions occur late, together with the element-wise comparisons. They may not be evaluated.
+        /// </summary>
+        public override BoundNode VisitTupleBinaryOperator(BoundTupleBinaryOperator node)
+        {
+            var boolType = node.Type; // we can re-use the bool type
+            var initialEffectsAndTemps = TupleOperatorSideEffectsAndTemps.GetInstance();
+
+            BoundExpression newLeft = ReplaceTerminalElementsWithTemps(node.Left, node.Operators, initialEffectsAndTemps.leftInit, initialEffectsAndTemps.temps);
+            BoundExpression newRight = ReplaceTerminalElementsWithTemps(node.Right, node.Operators, initialEffectsAndTemps.rightInit, initialEffectsAndTemps.temps);
+
+            var returnValue = RewriteTupleNestedOperators(node.Operators, newLeft, newRight, boolType, initialEffectsAndTemps, node.OperatorKind);
+
+            var (effects, temps) = initialEffectsAndTemps.ToImmutableAndFree();
+            var result = _factory.Sequence(temps, effects, returnValue);
```

Comment:
>var result = _factory.Sequence(temps, effects, returnValue); [](start = 12, length = 60)

Could there be a case when there are no temps and no side-effects? Are we going to create a sequence for this case anyway?